doc/source/development/rfc/rfc104_gdal_cli.rst

```diff
@@ -0,0 +1,1175 @@
+
+The GDALAlgorithm class itself has logic, triggered during the validation phase,
+to open the ``GDALDataset*`` from its name for input arguments, taking into
+account potential open options and allowed input formats.
+It has also very specific logic to realize that if both the ``input`` and
+``output`` arguments point to the same dataset name, a single ``GDALDataset*``
+instance must be set onto both for drivers that require it (typically SQLite
+based ones).
+
+GDALAlgorithmRegistry
+*********************
+
+Instances of this class store a list of C++ types implementing GDALAlgorithm.
+
+It is used by GDALAlgorithm itself for nodes to reference their potential children.
+
+It is extended by a GDALGlobalAlgorithmRegistry to offer a singleton that lists all
+top-level nodes (``raster``, ``vector``, etc.). Potentially code external to GDAL
+could register a new command available for use by :program:`gdal` in a GDAL plugin.
+
+It is also used by the ``GDALVectorPipelineAlgorithm`` to list its potential steps.
+
+C API
++++++
+
+Details will be fleshed out in further developments of the candidate implementation.
+
+This will at least includes:
+
+- a ``GDALAlgorithmH`` opaque handle to ``GDALAlgorithm``
+
+- Mapping main public methods of ``GDALAlgorithm`` as C functions, in
+  particular a ``bool GDALAlgorithmRun(GDALAlgorithmH, GDALProgressFunc, void* pProgressData)``
+
+- Mapping access to setting argument values.
+
+- Mapping access to the GDALGlobalAlgorithmRegistry.
+
+SWIG API
+++++++++
+
+Part of this C API will be made available in particular for the Python bindings,
+at least, for the sake of our Python autotest suite, as most of the testing
+will be done through that way.
+
+``gdal`` binary
++++++++++++++++
+
+gdal.cpp is a ~ 50 line of code launcher script that queries the ``gdal`` main
+algorithm, passes it to it the command line arguments and execute the
+``GDALAlgorithm::Run`` method.
+
+Out of scope
+------------
+
+* This RFC only addresses existing C++ utilities. Python utilities that would be
+  migrated in the future as C++ utilities should follow this RFC.
+
+* The very specific :ref:`sozip` utility will not follow this RFC. It has been
+  design to mimic the existing standard ``zip`` utility.
+
+Backward compatibility
+----------------------
+
+Fully backwards compatible. Existing utilities will remain for now, and if the
+project decides to retire them in the future, that will likely go through a
+multi-year deprecation period. Such decision will be made later, depending on
+the maturity of the new unified CLI approach, and its adoption status by the
+community.
+
+Before GDAL 3.11 release, we'll need to decide if we advertise the already
+implemented commands as stable or experimental. My feeling is that it might be
+prudent to label them as experimental for now, but release them as part of
+3.11.0 to get broader feedback from users, before stabilizing command and option
+names.
```

Comment:
I agree on being prudent. This is a big development with a big API surface (as CLI in this case).